README.md
# Protobuffy

(the JSON slayer)

[![Gem Version](https://badge.fury.io/rb/protobuf.svg)](http://badge.fury.io/rb/protobuf)
[![Build Status](https://secure.travis-ci.org/ruby-protobuf/protobuf.svg?branch=master)](https://travis-ci.org/ruby-protobuf/protobuf)
[![Gitter chat](https://badges.gitter.im/ruby-protobuf/protobuf.svg)](https://gitter.im/ruby-protobuf/protobuf)
[![protobuf API Documentation](https://www.omniref.com/ruby/gems/protobuf.png)](https://www.omniref.com/ruby/gems/protobuf)

Protobuffy is an implementation of [Google's protocol buffers][google-pb] in
ruby, version 2.5.0 is currently supported.

## Install

See our [Installation Guide][] on the [wiki][].

## Usage

The [wiki][] contains in-depth guides on the various ways to use this gem
including [compiling definitions][], [object APIs][], [services][], [clients][], and even
an [API roadmap][].

## Examples

In two different terminals run the follow pairs of commands:

Socket:

    bundle exec bin/rpc_server start --socket ./examples/lib/example/reverse-service.rb
    bundle exec examples/bin/reverse-client-socket '!skrow tekcos'

ZeroMQ:

    bundle exec bin/rpc_server start --zmq ./examples/lib/example/reverse-service.rb
    bundle exec examples/bin/reverse-client-zmq '!skrow qmorez'

HTTP:

    bundle exec bin/rpc_server start --http ./examples/lib/example/reverse-service.rb
    bundle exec examples/bin/reverse-client-http '!skrow ptth'

Alternatively, start the server as a Rack app using a `rackup` and a standard `config.ru`:

    bundle exec rackup examples/config.ru -p 9399

## Changelog

See recent changes in the [release notes][] or the [changelog][].

  [google-pb]:             http://code.google.com/p/protobuf "Google Protocol Buffers"
  [wiki]:                  https://github.com/ruby-protobuf/protobuf/wiki "Wiki home page"
  [Installation Guide]:    https://github.com/ruby-protobuf/protobuf/wiki/Installation "Installation guide"
  [compiling definitions]: https://github.com/ruby-protobuf/protobuf/wiki/Compiling-Definitions "Compiling guide"
  [object APIs]:           https://github.com/ruby-protobuf/protobuf/wiki/Messages-&-Enums "Message & Enum object APIs guide"
  [services]:              https://github.com/ruby-protobuf/protobuf/wiki/Services "Services object API guide"
  [clients]:               https://github.com/ruby-protobuf/protobuf/wiki/Clients "Client object API guide"
  [API roadmap]:           https://github.com/ruby-protobuf/protobuf/wiki/API-Roadmap "API Roadmap guide"
  [release notes]:         https://github.com/ruby-protobuf/protobuf/releases "Release notes"
  [changelog]:             https://github.com/ruby-protobuf/protobuf/blob/master/CHANGES.md "CHANGES.md"


CHANGES.md
# Stable (3.x)

4.0.0
-----------
- Merge in ruby-protobuf/protobuf

Upstream 3.7.0 (pre2)
-----------
- BUG: Track if a repeated field has been deliberately set (#325)

Upstream 3.7.0 (pre1)
-----------
- BUG: Revert to old behavior for setting repeated fields to nil
- BUG: Set binmode for protoc-gen-ruby STDIN and STDOUT to compile proto files on Windows
- Make all things Optionable and fix requires

Upstream 3.7.0 (pre0)
-----------
- Add `PB_USE_RAW_RPC_NAMES` option to preserve raw RPC name (since #underscore can be lossy).
- Add `PB_ENUM_UPCASE` option to generate enum values as upcased.
- Clean up dynamic code generation in prep for extension namespacing.
- Namespace extension fields.
- Field values should be stored via their fully qualified names
- Refresh google/protobuf/descriptor.{proto,pb.rb}
- Properly encode and decode negative enum values.

Upstream 3.6.9
--------
- Make protobuf serivce directory pluggable.

Upstream 3.6.7
-----
- An issue was reported with the encode memoization added in #293 with using any array modification
method on repeated fields. Remove memoization on encode (#305) until we can find a better solution.

Upstream 3.5.5
--------
- Add native Varint for MRI.

Upstream 3.5.4
--------
- Ensures ActiveSupport::Deprecation does not get a stack trace when deprecations are disabled.

Upstream 3.5.3
--------
- Optimized get_extension_field and get_field calls.

Upstream 3.5.2
--------
- Optimized valid_tag?, enums_for_tag and enums_for_tags

Upstream 3.5.1
--------
- Adds compatibility for Rails 4.2+ as CLI options were broken
- Fixes bug with MRI and "dead" thread in zmq broker
- Fixes Rubocop compatability with new version

3.2.0
--------

- Change Protobuf::Rpc::Stats.statsd_client to Protobuf::Statsd.client
    and track both RPC calls and service handling. [@dgolombek]

3.1.0
--------

- Add Protobuf::Rpc::Stats.statsd_client API to pass statistics on to Statsd. [@dgolombek]

3.0.4
--------

- Raise specific MethodNotFound when service class doesn't respond to (publicly implement)
    the rpc method called from the client. Stop rescuing all NoMethodError's thrown
    by service implementations. [#193, @liveh2o]

3.0.3
---------

- Fix recursive memory/cpu growth issue when calling class-level `Message.to_json`. [#190]

3.0.2
---------

- Queue requests at the broker when concurrent requests hit the ZMQ server, distribute to
    worker threads on each turn of the read poll loop. [#189, @abrandoned, @liveh2o]

3.0.1
---------

- Fix NoMethodError that can occur when serializing a message with a missing required field. [#187, @abrandoned]

3.0.0
---------

A lot has changed since the last stable v2.8.12. For all the relevant changes,
see the closed [pull requests and issues list in github](https://github.com/localshred/protobuf/issues?milestone=1&state=closed).
Below is a high-level list of fixes, deprecations, breaking changes, and new APIs.

### EventMachine is dead, Long live EventMachine

The EventMachine client and server have been removed from this gem. They code was far
too error prone and flawed to feasibly support going forward. It's recommended
to switch to the socket implementation (using `PB_CLIENT_TYPE` and `PB_SERVER_TYPE` of `socket`)
for a painless switchover. The ZMQ implementation is much more performant but
does have a dependency on libzmq.

### Server Middlewares

The server/dispatcher stack has been converted to the Middleware pattern!
Exception handling (#162, #164), Request decoding (#160, #166), Response encoding (#161, #167),
Logging and stats (#163), and Method dispatch (#159) have all been extracted into their
own Middlewares to greatly simplify testing and further development of the server
stack, furthering our preparations for removing the socket implementations (zmq, socket)
into their own gems in version 4.0. Major props to @liveh2o for [tackling this beast](https://github.com/localshred/protobuf/tree/master/lib/protobuf/rpc/middleware).

#### Bug Fixes

- Resolve DNS names (e.g. localhost) when using ZMQ server. [#46, reported by @reddshack]
- Switched to hash based value storage for messages to fix large field tag memory issues. [#118, #165]
- `Enum.fetch` used to return an enum of any type if that is the value passed in. [#168]

#### Deprecations

__!! NOTE: These deprecated methods will be removed in v3.1. !!__

- Deprecated `BaseField#type` in favor of `#type_class`.
- Deprecated `Message.get_ext_field_by_name` in favor of `.get_extension_field` or `.get_field(name_or_tag, true)`.
- Deprecated `Message.get_ext_field_by_tag,` in favor of `.get_extension_field` or `.get_field(name_or_tag, true)`.
- Deprecated `Message.get_field_by_name,` in favor of `.get_field`.
- Deprecated `Message.get_field_by_tag,` in favor of `.get_field`.
- Deprecated `Enum.enum_by_value` in favor of `.enum_for_tag`.
- Deprecated `Enum.name_by_value` in favor of `.name_for_tag`.
- Deprecated `Enum.get_name_by_tag` in favor of `.name_for_tag`.
- Deprecated `Enum.value_by_name` in favor of `.enum_for_name`.
- Deprecated `Enum.values` in favor of `.enums`. Beware that `.enums` returns an array where `.values` returns a hash.
   Use `.all_tags` if you just need all the valid tag numbers. In other words, don't do this anymore: `MyEnum.values.values.map(&:to_i).uniq`.

#### Breaking Changes

- Require Active Support 3.2+. [#177]
- All files/classes relating to the EventMachine client and server are gone. Use `PB_CLIENT_TYPE` and `PB_SERVER_TYPE` of `socket` instead. [#116]
- Cleaned up the `Enum` class, deprecating/renaming most methods. tl;dr, just use `MyEnum.fetch`.
   See #134 for more comprehensive documentation about which methods are going and away and which are being renamed. [#134]
- Pulled `EnumValue` into `Enum`. The `EnumValue` class no longer exists. Use `Enum` for type-checking instead. [#168].
- Removed previously deprecated `bin/rprotoc` executable. Use `protoc --ruby_out=...` instead. [13fbdb9]

lib/protobuf/cli.rb
require 'active_support/core_ext/hash/keys'
require 'active_support/inflector'

require 'thor'
require 'protobuf/version'
require 'protobuf/logging'
require 'protobuf/rpc/servers/socket_runner'
require 'protobuf/rpc/servers/zmq_runner'
require 'protobuf/rpc/servers/http_runner'

module Protobuf
  class CLI < ::Thor
    include ::Thor::Actions
    include ::Protobuf::Logging

    attr_accessor :runner, :mode, :exit_requested

    no_commands do
      alias_method :exit_requested?, :exit_requested
    end

    default_task :start

    desc 'start APP_FILE', 'Run the RPC server in the given mode, preloading the given APP_FILE. This is the default task.'

    option :host,                       :type => :string, :default => '127.0.0.1', :aliases => %w(-o), :desc => 'Host to bind.'
    option :port,                       :type => :numeric, :default => 9399, :aliases => %w(-p), :desc => 'Master Port to bind.'

    option :backlog,                    :type => :numeric, :default => 100, :aliases => %w(-b), :desc => 'Backlog for listening socket when using Socket Server.'
    option :threshold,                  :type => :numeric, :default => 100, :aliases => %w(-t), :desc => 'Multi-threaded Socket Server cleanup threshold.'
    option :threads,                    :type => :numeric, :default => 5, :aliases => %w(-r), :desc => 'Number of worker threads to run. Only applicable in --zmq mode.'

    option :log,                        :type => :string, :default => STDOUT, :aliases => %w(-l), :desc => 'Log file or device. Default is STDOUT.'
    option :level,                      :type => :numeric, :default => ::Logger::INFO, :aliases => %w(-v), :desc => 'Log level to use, 0-5 (see http://www.ruby-doc.org/stdlib/libdoc/logger/rdoc/)'

    option :socket,                     :type => :boolean, :aliases => %w(-s), :desc => 'Socket Mode for server and client connections.'
    option :zmq,                        :type => :boolean, :aliases => %w(-z), :desc => 'ZeroMQ Socket Mode for server and client connections.'
    option :http,                       :type => :boolean, :aliases => %w(), :desc => 'HTTP Server Mode for server and client connections.'

    option :beacon_interval,            :type => :numeric, :desc => 'Broadcast beacons every N seconds. (default: 5)'
    option :beacon_port,                :type => :numeric, :desc => 'Broadcast beacons to this port (default: value of ServiceDirectory.port)'
    option :broadcast_beacons,          :type => :boolean, :desc => 'Broadcast beacons for dynamic discovery (Currently only available with ZeroMQ).'
    option :broadcast_busy,             :type => :boolean, :default => false, :desc => 'Remove busy nodes from cluster when all workers are busy (Currently only available with ZeroMQ).'
    option :debug,                      :type => :boolean, :default => false, :aliases => %w(-d), :desc => 'Debug Mode. Override log level to DEBUG.'
    option :gc_pause_request,           :type => :boolean, :default => false, :desc => 'DEPRECATED: Enable/Disable GC pause during request.'
    option :print_deprecation_warnings, :type => :boolean, :default => nil, :desc => 'Cause use of deprecated fields to be printed or ignored.'
    option :workers_only,               :type => :boolean, :default => false, :desc => "Starts process with only workers (no broker/frontend is started) only relevant for Zmq Server"
    option :worker_port,                :type => :numeric, :default => nil, :desc => "Port for 'backend' where workers connect (defaults to port + 1)"
    option :zmq_inproc,                 :type => :boolean, :default => true, :desc => 'Use inproc protocol for zmq Server/Broker/Worker'

    def start(app_file)
      debug_say('Configuring the rpc_server process')

      configure_logger
      configure_traps
      configure_runner_mode
      create_runner
      configure_process_name(app_file)
      configure_gc
      configure_deprecation_warnings

      require_application(app_file) unless exit_requested?
      start_server unless exit_requested?
    rescue => e
      say_and_exit('ERROR: RPC Server failed to start.', e)
    end

    desc 'version', 'Print ruby and protoc versions and exit.'
    def version
      say("Ruby Protobuf v#{::Protobuf::VERSION}")
    end

    no_tasks do

      # Tell protobuf how to handle the printing of deprecated field usage.
      def configure_deprecation_warnings
        ::Protobuf.print_deprecation_warnings =
          if options.print_deprecation_warnings.nil?
            !ENV.key?("PB_IGNORE_DEPRECATIONS")
          else
            options.print_deprecation_warnings?
          end
      end

      # If we pause during request we don't need to pause in serialization
      def configure_gc
        say "DEPRECATED: The gc_pause_request option is deprecated and will be removed in 4.0." if options.gc_pause_request?

        debug_say('Configuring gc')

        ::Protobuf.gc_pause_server_request =
          if defined?(JRUBY_VERSION)
            # GC.enable/disable are noop's on Jruby
            false
          else
            options.gc_pause_request?
          end
      end

      # Setup the protobuf logger.
      def configure_logger
        debug_say('Configuring logger')

        log_level = options.debug? ? ::Logger::DEBUG : options.level

        ::Protobuf::Logging.initialize_logger(options.log, log_level)

        # Debug output the server options to the log file.
        logger.debug { 'Debugging options:' }
        logger.debug { options.inspect }
      end

      # Re-write the $0 var to have a nice process name in ps.
      def configure_process_name(app_file)
        debug_say('Configuring process name')
        $0 = "rpc_server --#{mode} #{options.host}:#{options.port} #{app_file}"
      end

      # Configure the mode of the server and the runner class.
      def configure_runner_mode
        debug_say('Configuring runner mode')
        server_type = ENV["PB_SERVER_TYPE"]

        self.mode = if multi_mode?
                      say('WARNING: You have provided multiple mode options. Defaulting to socket mode.', :yellow)
                      :socket
                    elsif options.zmq?
                      :zmq
                    elsif options.http?
                      :http
                    else
                      case server_type
                      when nil, /\Asocket[[:space:]]*\z/i
                        :socket
                      when /\Azmq[[:space:]]*\z/i
                        :zmq
                      when /http/i
                        :http
                      else
                        require server_type.to_s
                        server_type
                      end
                    end
      end

      # Configure signal traps.
      # TODO: add signal handling for hot-reloading the application.
      def configure_traps
        debug_say('Configuring traps')

        exit_signals = [:INT, :TERM]
        exit_signals << :QUIT unless defined?(JRUBY_VERSION)

        exit_signals.each do |signal|
          debug_say("Registering trap for exit signal #{signal}", :blue)

          trap(signal) do
            self.exit_requested = true
            shutdown_server
          end
        end
      end

      # Create the runner for the configured mode
      def create_runner
        debug_say("Creating #{mode} runner")
        self.runner = case mode
                      when :zmq
                        create_zmq_runner
                      when :socket
                        create_socket_runner
                      when :http
                        create_http_runner
                      else
                        say("Extension runner mode: #{mode}")
                        create_extension_server_runner
                      end
      end

      # Say something if we're in debug mode.
      def debug_say(message, color = :yellow)
        say(message, color) if options.debug?
      end

      # Internal helper to determine if the modes are multi-set which is not valid.
      def multi_mode?
        options.zmq? && options.socket?
      end

      # Require the application file given, exiting if the file doesn't exist.
      def require_application(app_file)
        debug_say('Requiring app file')
        require app_file
      rescue LoadError => e
        say_and_exit("Failed to load application file #{app_file}", e)
      end

      def runner_options
        opt = options.to_hash.symbolize_keys

        opt[:workers_only] = (!!ENV['PB_WORKERS_ONLY']) || options.workers_only

        opt
      end

      def say_and_exit(message, exception = nil)
        message = set_color(message, :red) if options.log == STDOUT

        logger.error { message }

        if exception
          $stderr.puts "[#{exception.class.name}] #{exception.message}"
          $stderr.puts exception.backtrace.join("\n")

          logger.error { "[#{exception.class.name}] #{exception.message}" }
          logger.debug { exception.backtrace.join("\n") }
        end

        exit(1)
      end

      def create_extension_server_runner
        classified = mode.classify
        extension_server_class = classified.constantize

        self.runner = extension_server_class.new(runner_options)
      end

      def create_socket_runner
        require 'protobuf/socket'

        self.runner = ::Protobuf::Rpc::SocketRunner.new(runner_options)
      end

      def create_zmq_runner
        require 'protobuf/zmq'

        self.runner = ::Protobuf::Rpc::ZmqRunner.new(runner_options)
      end

      def create_http_runner
        require 'protobuf/http'

        self.runner = ::Protobuf::Rpc::HttpRunner.new(runner_options)
      end

      def shutdown_server
        logger.info { 'RPC Server shutting down...' }
        runner.stop
        ::Protobuf::Rpc::ServiceDirectory.instance.stop
      end

      # Start the runner and log the relevant options.
      def start_server
        debug_say('Running server')

        runner.run do
          logger.info do
            "pid #{::Process.pid} -- #{mode} RPC Server listening at #{options.host}:#{options.port}"
          end

          ::ActiveSupport::Notifications.instrument("after_server_bind")
        end

        logger.info { 'Shutdown complete' }
      end
    end
  end
end
